init: Add option for rpccookie permissions (replace 26088)

willcl-ark opened the pull request:
This PR picks up #26088 by aureleoules which adds a bitcoind launch option `-rpccookieperms` to set the file permissions of the cookie generated by bitcoin core.

Example usage to make the generated cookie group-readable: `./src/bitcoind -rpccookieperms=group`.

Accepted values for `-rpccookieperms` are `[owner|group|all]`. We let `fs::perms` handle platform-specific permissions changes.

luke-jr commented on src/httprpc.cpp:
```diff
@@ -239,12 +239,37 @@ static bool HTTPReq_JSONRPC(const std::any& context, HTTPRequest* req)
 
+static std::optional<unsigned> StringToOctal(const std::string& str)
+{
+    if (str.length() != 4) return std::nullopt;
```
Not sure `StringToOctal` is the right place for this check.

luke-jr commented on src/httprpc.cpp:
```diff
@@ -239,12 +239,37 @@ static bool HTTPReq_JSONRPC(const std::any& context, HTTPRequest* req)
 
+static std::optional<unsigned> StringToOctal(const std::string& str)
+{
+    if (str.length() != 4) return std::nullopt;
```
Isn't the leading 0 optional? (Also the user might want to setgid or something weird)

willcl-ark replied:
b4b0d2adc9 now handles optional special mode bit

luke-jr commented on src/httprpc.cpp:
```diff
@@ -239,12 +239,43 @@ static bool HTTPReq_JSONRPC(const std::any& context, HTTPRequest* req)
+static std::optional<unsigned> ConvertPermsToOctal(const std::string& str)
+{
+    // Padding the front with 0 sets no special modes if not provided
+    if (str.length() == 3) return StringToOctal("0" + str);
```
I don't think we need to prepend the zero here.

willcl-ark replied:
Ok removed the 0 padding in 1b0fa12fa5

willcl-ark commented:
Updated to address luke's comment and added a test to ensure rpccookieperms are being applied

willcl-ark commented:
Rebased now that #28905 is merged.

luke-jr commented on src/rpc/request.cpp:
```diff
@@ -82,7 +82,26 @@ static fs::path GetAuthCookieFile(bool temp=false)
+static std::string perms_to_str(fs::perms p)
+{
+    char special = '-';
+    if ((p & fs::perms::set_uid)    != fs::perms::none) special = '4';
+    if ((p & fs::perms::set_gid)    != fs::perms::none) special = '2';
+    if ((p & fs::perms::sticky_bit) != fs::perms::none) special = '1';
```
This won't work if multiple bits are set

luke-jr replied:
These bits seem to typically manifest by replacing the 'x' field with either 's' (setuid/setgid+executable), 'S' (setuid/setgid WITHOUT executable), 't' (sticky+exec), or 'T' (sticky NO exec)

kristapsk commented:
Should the permissions that can be set limited here? There is no reason to ever set +x on an RPC cookie file.

luke-jr commented:
I'm not sure there's a reason to ever set +w either... maybe it should just be -rpccookieaccess user/group/all or something

kristapsk commented:
> I'm not sure there's a reason to ever set +w either... maybe it should just be -rpccookieaccess user/group/all or something

Agree, makes sense. Giving read access to other users is the only use case of this functionality I know of.

willcl-ark commented:
Sorry for the slow response. I was away buit am now back.

Thanks for the review, I agree that setting special bits is uninteresting for cookie files (I was really trying to keep the subfunctions generic in case we wanted to re-use them elsewhere). I've now removed ability to set special bits.

I think a little on switching to string-based settings, i.e. `-rpccookieperms=group`, but this quickly gets more ugly inside, as IMO the usual use-case is to be setting multiple values in the same permission. i.e. setting just `user` (`040`) means that the owner-user will NOT be able to read the cookie, as user is checked against owner before group, and if a match is found, with no read perms, the checks are exited. I feel that users are likely to want to set `group` in _addition_ to `user`, therefore 3 digit ocal perms make more sense.

On a related note, I did wonder whether I should introduce a preliminary commit which sets our default cookie perms to `400`, rather than the `600` they get from the umask, but

kristapsk commented:
> I think a little on switching to string-based settings, i.e. `-rpccookieperms=group`, but this quickly gets more ugly inside, as IMO the usual use-case is to be setting multiple values in the same permission. i.e. setting just `user` (`040`) means that the owner-user will NOT be able to read the cookie, as user is checked against owner before group, and if a match is found, with no read perms, the checks are exited. I feel that users are likely to want to set `group` in _addition_ to `user`, therefore 3 digit ocal perms make more sense.

Only three permissions that makes sense IMO are `600`, `640` and `644`.

willcl-ark commented:
> > I think a little on switching to string-based settings, i.e. `-rpccookieperms=group`, but this quickly gets more ugly inside, as IMO the usual use-case is to be setting multiple values in the same permission. i.e. setting just `user` (`040`) means that the owner-user will NOT be able to read the cookie, as user is checked against owner before group, and if a match is found, with no read perms, the checks are exited. I feel that users are likely to want to set `group` in _addition_ to `user`, therefore 3 digit ocal perms make more sense.
> 
> Only three permissions that makes sense IMO are `600`, `640` and `644`.

Yes. I did author a commit to be prepended the two in this PR, which changed our default cookie perms to `400` using `DEFAULT_COOKIE_PERMS` (with current umask it is `600`), as I don't think enabling write perms on a cookie (ever?!) that bitcoind has created makes sense? It just seemed a bit out of scope for this PR and I thought it might work best as a followup (or before) on it's own. It could also include rejecting permissions which included writing if we wanted...

willcl-ark committed:
init: add option for rpccookie permissions

Add a bitcoind launch option `-rpccookieperms` to configure the file
permissions of the cookie on Unix systems.

willcl-ark committed:
test: add rpccookieperms test

Tests various perms on non-Windows OSes

willcl-ark commented:
OK I added a commit to change default cookie perms to `400`, and rebased on master to try and get the unrelated CI failure to go away.

jamesob commented:
Approach ACK; seems like a reasonable feature reasonably implemented. Will do a more formal review soon.

luke-jr commented on src/httprpc.cpp:
```diff
@@ -240,12 +240,41 @@ static bool HTTPReq_JSONRPC(const std::any& context, HTTPRequest* req)
         LogPrintf("Using random cookie authentication.\n");
-        if (!GenerateAuthCookie(&strRPCUserColonPass)) {
+
+        auto cookie_perms{DEFAULT_COOKIE_PERMS};
```
I dislike using `auto` both here and the default. It could conceivably not be a type that supports all permission masks.

willcl-ark replied:
Changed to `fs::perms` here and the default in fcd100cb8e35e7cb00a58a697274a1f2cdb3cc75

luke-jr commented on src/rpc/request.cpp:
```diff
@@ -82,23 +82,32 @@ static fs::path GetAuthCookieFile(bool temp=false)
     }
+
+#ifdef WIN32
+    LogPrintf("Unable to set unix file permissions on cookie using Windows systems: %s\n", fs::PathToString(filepath_tmp));
```
Feels weird to log this unconditionally. Maybe the option should just be unavailable (and implicitly error) on Windows, and hard-code setting it user-read-only here?

willcl-ark replied:
The cookie should be read-only on Windows by deafult, so I've moved the logging up into `InitRPCAuthentication` [here](https://github.com/bitcoin/bitcoin/pull/28167/commits/fcd100cb8e35e7cb00a58a697274a1f2cdb3cc75#diff-a8ffd831b801d5170ec47bd2cde728ada0ffb3571f03ad70d93fc66c2009309aR269-R272), which now also returns an error if the user tries to set `rpccookieperms`  on windows.

luke-jr replied:
I don't think Windows has strict defaults on permissions for new files, with regard to other users. [This StackExchange article](https://superuser.com/questions/242783/set-default-access-control-list-acl-to-everyone-when-creating-a-new-file-or-fo) at least suggests the default can be manipulated manually.

Also, typically when a feature is disabled at build time, we don't add it to argsman (except perhaps as a hidden_args, but I think that may be a bug in such cases).

luke-jr commented on test/functional/rpc_users.py:
```diff
@@ -84,6 +94,39 @@ def test_auth(self, node, user, password):
+            actual_perms = f"{(file_stat.st_mode & PERM_BITS_UMASK):03o}"
+            assert_equal(perm, actual_perms)
+
+       # Remove any leftover rpc{user|password} config options from previous tests
```
Missing a space in the indentation

willcl-ark replied:
Thanks! My formatter is having some difficulty with diff/range-based formatting at the moment. Fixed.

luke-jr commented on test/functional/rpc_users.py:
```diff
@@ -84,6 +94,39 @@ def test_auth(self, node, user, password):
+        test_perm(None)
+
+        self.log.info('Check custom cookie permissions')
+        test_perms = ["440", "640", "444"]
+        [test_perm(perm) for perm in test_perms]
```
Feels like a hack

```suggestion
        for perm in "440", "640", "444":
            test_perm(perm)
```

willcl-ark replied:
IMO list comprehensions are fine, but seeing as I voided half the point of using one (creating a one-liner), so took the change to a regular for loop in 7516e9120e202395dabebd5b7212d9c2d3787c43

luke-jr commented:
Actually, why wouldn't fs::permissions work on Windows?

luke-jr commented:
For your consideration, I have pushed a variant with some recommended changes here: https://github.com/luke-jr/bitcoin/commits/rpccookieperms-26%2Bknots/

Feel free to cherry-pick or squash into your PR as you deem best.

willcl-ark commented:
Thanks Luke, I agree that repeated string concatenation was not a particularly neat for creating a single string where we could simply modify in place. I've updated this further from your version to use a lambda in bf9b11d8f6c0af9b2c76674481e5905a00240cf5.

Re Windows. Yes, fs::permissions is "supported", but IIUC without devolving into access control lists, it all boils down to whether the write bit is enabled. I don't see the sense in supporting either of these things here personally. The default cookie permission should be setting the read flag on Windows, but not write. And I would prefer to keep `rpccookieperms` Unix only.

tdb3 reviewed (commented):
Test ACK ce9df2aba3e98ba7f2a576d587ba8e59bf341083.

Great and appreciated addition.

Performed basic manual testing, receiving all successful/expected results.

1. Confirmed default of 400 perms for the `.cookie` file when no CLI args or config file parameters are provided.
1. Confirmed CLI arg of `-rpccookieperms=440` results in 440 permissions.
1. Confirmed bitcoin.conf with `rpccookieperms=440` results in 440 permissions.
1. Confirmed empty, non-number, and > 3 octal number is detected and error provided (expected).

nit: Recommend augmenting content in doc/init.md to inform of default cookie file permissions.
Example in this commit https://github.com/tdb3/bitcoin/commit/fdd8a4e70632c74c630caaff5cc36da0c309168e (welcome to cherry pick as desired).
`bitcoind --help` already states default permissions of 400, so the addition to init.md is author's discretion.

willcl-ark commented:
> Dealing with a read-only tmp file is probably outside the scope of this PR, but maybe we should reconsider if +rw might actually be the correct mode to use.

@luke-jr that seems like an unlikely situation (unless bitcoind is killed in the moment between writing the temporary cookie file and renaming it?), nonetheless, it could be addressed by changing where we apply the new more restrictive `400` permissions, from the temp file to the renamed `.cookie`.

I don't think the temporary cookie file existing with perms `600` for a moment is particularly problematic, so I've integrated the change here in 06761eb2686ffe3d7f647d8df9bb7b555bc0072a

I also tested that Bitcoin Core starts up OK like this with the new changes:

```bash
cd $DATADIR
touch .cookie.tmp
chmod 600 .cookie.tmp
./src/bitcoind
```

Even if the `.cookie.tmp` file contains old data from a previous start, the default `std::ofstream` open mode will overwrite the file rather than append, so after renaming the `.cookie` should always be valid.

luke-jr commented:
My point was more that we should probably not be making cookie files read-only to begin with. So only `rw-r--r--`, `rw-r-----`, and `rw-------` should be supported. Bringing us back to this maybe being best as `-rpccookieperms=user/group/all` rather than an octal modeset.

And then how to handle if the cookie file already exists read-only... it *might* make sense to just use it unmodified in that scenario?

willcl-ark committed:
doc: detail -rpccookieperms option

Co-authored-by: tdb3 <106488469+tdb3@users.noreply.github.com>

willcl-ark commented:
OK Luke you've persuaded me. I've re-worked this now to accept `-rpccookieperms=[user|group|others]`.

> nit: Recommend augmenting content in doc/init.md to inform of default cookie file permissions.
Example in this commit https://github.com/tdb3/bitcoin/commit/fdd8a4e70632c74c630caaff5cc36da0c309168e (welcome to cherry pick as desired).
bitcoind --help already states default permissions of 400, so the addition to init.md is author's discretion.

Thansk @tdb3, I've taken your suggestion in b02736b2fb0ec2ac1ac9d6100e7677348394d306

luke-jr commented on src/init.cpp:
```diff
@@ -650,6 +650,7 @@ void SetupServerArgs(ArgsManager& argsman)
     argsman.AddArg("-rpcbind=<addr>[:port]", "Bind to given address to listen for JSON-RPC connections. Do not expose the RPC server to untrusted networks such as the public internet! This option is ignored unless -rpcallowip is also passed. Port is optional and overrides -rpcport. Use [host]:port notation for IPv6. This option can be specified multiple times (default: 127.0.0.1 and ::1 i.e., loc ...
     argsman.AddArg("-rpcdoccheck", strprintf("Throw a non-fatal error at runtime if the documentation for an RPC is incorrect (default: %u)", DEFAULT_RPC_DOC_CHECK), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::RPC);
     argsman.AddArg("-rpccookiefile=<loc>", "Location of the auth cookie. Relative paths will be prefixed by a net-specific datadir location. (default: data dir)", ArgsManager::ALLOW_ANY, OptionsCategory::RPC);
+    argsman.AddArg("-rpccookieperms=<readable-by>", strprintf("Set permissions on the RPC auth cookie file so that it is readable by [owner|group|all] (default: %u)", DEFAULT_COOKIE_PERMS.name), ArgsManager::ALLOW_ANY, OptionsCategory::RPC);
```
`%s`

willcl-ark replied:
Fixed in 5a20b078c3a418103601ee6eb88e5dba8e6d9863

luke-jr commented on src/util/fs_helpers.h:
```diff
@@ -62,6 +62,19 @@ void ReleaseDirectoryLocks();
 
+std::string PermsToString(fs::perms p);
+
+struct CookiePerms {
+    struct Perm {
+        const char* name;
+        fs::perms permissions;
+    };
+
+    static constexpr Perm owner  {"owner",  fs::perms::owner_read | fs::perms::owner_write};
+    static constexpr Perm group  {"group",  owner.permissions     | fs::perms::group_read};
+    static constexpr Perm all    {"all",    group.permissions     | fs::perms::others_read};
+};
```
This feels overengineered. Why can't we stick to fs::perms internally and just map the string in a function?

willcl-ark replied:
Now simplified in d8d47ab395b8fa341e0aac2fe01a4b9488af5347

willcl-ark commented on src/util/fs_helpers.cpp:
```diff
@@ -269,3 +270,42 @@ bool TryCreateDirectories(const fs::path& p)
+    } else if (s == "all") {
+        return fs::perms::owner_read | fs::perms::owner_write |
+               fs::perms::group_read | fs::perms::group_write |
+               fs::perms::others_read | fs::perms::others_write;
```
Reverted to RO for group and others. I had misunderstood what you meant here:

> My point was more that we should probably not be making cookie files read-only to begin with. So only rw-r--r--, rw-r-----, and rw------- should be supported. Bringing us back to this maybe being best as -rpccookieperms=user/group/all rather than an octal modeset.

IMO it doesn't matter much if users and others get rw or ro, if they can read it, they can access the RPCs.

luke-jr commented:
QA failed

ryanofsky reviewed (approved):
Code review ACK 9617e42a7b12a05ed8a0815c5a6537e1614727f5. I didn't follow all the previous discussion, but I like the place this PR ended up accepting simple "owner" "group" and "all" settings and avoiding more complicated corner cases.

I left some suggestions which are mostly not important, though I do think it would be better to try to stay more backwards compatible and avoid setting permissions when it is not requested.

tdb3 reviewed (commented):
Approach ACK.
This is looking great.

Left a comment in `test_rpccookieperms()`.

Also manually checked cookie file permissions after starting bitcoind without `-rpccookieperms` and with each option (`owner`, `group`, `all`) respectively.  Permissions were as expected:
```
-rw------- 1 dev dev   75 Jun  7 21:59 .cookie
```
```
-rw------- 1 dev dev   75 Jun  7 21:59 .cookie
```
```
-rw-r----- 1 dev dev   75 Jun  7 22:00 .cookie
``` 
```
-rw-r--r-- 1 dev dev   75 Jun  7 22:00 .cookie
```

willcl-ark commented on src/httprpc.cpp:
```diff
@@ -244,8 +246,25 @@ static bool InitRPCAuthentication()
+#ifdef WIN32
+            LogInfo("Unable to set unix-style file permissions on cookie via -rpccookieperms on Windows systems\n");
+            return false;
+#else
```
Thanks, now done in 7ad5349a2ec

willcl-ark commented on src/httprpc.h:
```diff
@@ -6,6 +6,12 @@
+/** Default permissions for cookie file.
+ * Set to owner read/write, which on Windows sets the write permission.
+ */
+const fs::perms DEFAULT_COOKIE_PERMS{fs::perms::owner_read | fs::perms::owner_write};
```
Dropped the new constant leaving the current behaviour as default, unless an optional `fs::perms` is provided in 7ad5349a2ec

willcl-ark commented on src/rpc/request.cpp:
```diff
@@ -82,33 +81,38 @@ static fs::path GetAuthCookieFile(bool temp=false)
 static bool g_generated_cookie = false;
 
-bool GenerateAuthCookie(std::string *cookie_out)
+bool GenerateAuthCookie(std::string* cookie_out, fs::perms cookie_perms)
```
Taken in 7ad5349a2ec

willcl-ark commented on test/functional/rpc_users.py:
```diff
@@ -84,6 +86,44 @@ def test_auth(self, node, user, password):
+            lines = file.readlines()
+        filtered_lines = [line for line in lines if not line.startswith('rpcuser') and not line.startswith('rpcpassword')]
+        with conf.open('w') as file:
+            file.writelines(filtered_lines)
```
Thanks, i took this simplification in 7ad5349a2ec

ryanofsky commented on src/rpc/request.cpp:
```diff
@@ -82,33 +81,40 @@ static fs::path GetAuthCookieFile(bool temp=false)
+        LogInfo("Unable to rename cookie authentication file %s to %s\n", fs::PathToString(filepath_tmp), fs::PathToString(filepath));
         return false;
     }
+    std::error_code code;
```
In commit "init: add option for rpccookie permissions" (7ad5349a2ec04399c99c8cbf71c8d3b4627132f8)

Could move this variable inside the if statement since it's not used afterwards.

willcl-ark replied:
Moved in 9eff5601059

ryanofsky reviewed (approved):
Code review ACK e4798038c00e787023b9dedc907966a08592d79f. Seems very clean now with the test simplification and without the win32 stuff.

tdb3 commented on src/rpc/request.h:
```diff
@@ -7,17 +7,19 @@
 
 /** Generate a new RPC authentication cookie and write it to disk */
-bool GenerateAuthCookie(std::string *cookie_out);
+bool GenerateAuthCookie(std::string* cookie_out, std::optional<fs::perms> cookie_perms);
```
nit:  If touching this file again, may want to consider defining a default argument for cookie_perms.  Since std::optional is used, I'm assuming the intent is not to force the caller to provide fs::perms.

```diff
diff --git a/src/rpc/request.h b/src/rpc/request.h
index b40df16631..c7e723d962 100644
--- a/src/rpc/request.h
+++ b/src/rpc/request.h
@@ -19,7 +19,7 @@ std::string JSONRPCReply(const UniValue& result, const UniValue& error, const Un
 UniValue JSONRPCError(int code, const std::string& message);
 
 /** Generate a new RPC authentication cookie and write it to disk */
-bool GenerateAuthCookie(std::string* cookie_out, std::optional<fs::perms> cookie_perms);
+bool GenerateAuthCookie(std::string* cookie_out, std::optional<fs::perms> cookie_perms=std::nullopt);
 /** Read the RPC authentication cookie from disk */
 bool GetAuthCookie(std::string *cookie_out);
 /** Delete RPC authentication cookie from disk */
```

`GenerateAuthCookie()` is currently used with the argument provided, so not a big deal either way.

willcl-ark replied:
Added in 49d5bfdd7e7

tdb3 reviewed (approved):
re ACK e4798038c00e787023b9dedc907966a08592d79f

Ran `rpc_users` locally (passed).
Re-ran the manual check in https://github.com/bitcoin/bitcoin/pull/28167#pullrequestreview-2105557076 (same results).
Left one minor nit.

achow101 commented on test/functional/rpc_users.py:
```diff
@@ -84,6 +86,39 @@ def test_auth(self, node, user, password):
+    def test_rpccookieperms(self):
+        p = {"owner": 0o600, "group": 0o640, "all": 0o644}
+
+        if os.name == 'nt':
```
In 31cc8bc305103baf2c9979d86ef1b43dec5628cb "test: add rpccookieperms test"

Since #29034, the preferred convention is to use `platform.system()` rather than `os.name`.

willcl-ark replied:
Thanks, updated in 49d5bfdd7e7

achow101 commented on test/functional/rpc_users.py:
```diff
@@ -84,6 +86,39 @@ def test_auth(self, node, user, password):
+        p = {"owner": 0o600, "group": 0o640, "all": 0o644}
+
+        if os.name == 'nt':
+            raise SkipTest(f"Skip cookie file permissions checks as OS detected as: {os.name=}")
```
In 31cc8bc305103baf2c9979d86ef1b43dec5628cb "test: add rpccookieperms test"

This is incorrect. It causes the entire test file to be marked as skipped (different exit code, marked differently in the test runner output) even though everything else ran. The correct thing here is to skip this individual case, which can be achieved by simply returning.

tdb3 replied:
Good catch.  Might also be good to print a log message before returning for awareness.

willcl-ark replied:
Oh no! fixed in 49d5bfdd7e7

achow101 commented on src/util/fs_helpers.h:
```diff
@@ -62,6 +63,9 @@ void ReleaseDirectoryLocks();
 bool TryCreateDirectories(const fs::path& p);
 fs::path GetDefaultDataDir();
 
+std::string PermsToString(fs::perms p);
+std::optional<fs::perms> StringToPerms(const std::string& s);
```
In 38af55e285336cdd3f42635938f24622451703b0 "util: add PermsToString and StringToPerms"

nit: This naming is a little bit confusing to me since it suggests that these functions are inverses of each other, but they're not as their string outputs are not compatible.

willcl-ark replied:
I agree on reflection, gave them new names in 49d5bfdd7e7

achow101 commented on src/rpc/request.cpp:
```diff
@@ -82,33 +81,40 @@ static fs::path GetAuthCookieFile(bool temp=false)
     GetRandBytes(rand_pwd);
     std::string cookie = COOKIEAUTH_USER + ":" + HexStr(rand_pwd);
 
-    /** the umask determines what permissions are used to create this file -
-     * these are set to 0077 in common/system.cpp.
```
In 7ad5349a2ec04399c99c8cbf71c8d3b4627132f8 "init: add option for rpccookie permissions"

This comment is removed, but the umask is still what sets the permissions in the default case.

Since we can set the permissions explicitly in this function, I think it would make sense to always set them with `owner` as default rather than relying on the umask that happens somewhere else.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/28167#discussion_r1655290944

> This comment is removed, but the umask is still what sets the permissions in the default case.

Good catch, it would be good to add back the comment.

> Since we can set the permissions explicitly in this function, I think it would make sense to always set them with `owner` as default rather than relying on the umask that happens somewhere else.

The PR was actually doing this originally, but I suggested doing the opposite in https://github.com/bitcoin/bitcoin/pull/28167#discussion_r1628110701. It makes sense to me that when `-rpccookieperms` option is not specified, bitcoind would not set permissions on this file, just like it is not setting permissions on other files.

It seems fragile to rely on the [fs::permissions](https://en.cppreference.com/w/cpp/filesystem/permissions) call doing the right thing on all filesystems, when its behavior is not specified anywhere and we have never relied on it before. The documentation points to some odd ways it can be behave like "changing some bits may automatically change others" and we don't know what unusual things it may do on fuse filesystems or filesystems that use ACLs.

willcl-ark replied:
I've un-removed the comment in 9eff5601059, as I agree with @ryanofsky that not changing the existant behaviour is the "safer" default, and new behaviour can be introduced with the new option.

willcl-ark committed:
util: add perm string helper functions

PermsToSymbolicString will convert from fs::perms to string type
'rwxrwxrwx'.

InterpretPermString will convert from a user-supplied "perm string" such
as 'owner', 'group' or 'all, into appropriate fs::perms.

tdb3 commented on src/util/fs_helpers.h:
```diff
@@ -62,6 +63,19 @@ void ReleaseDirectoryLocks();
+ * @param[in] s Permission level string
+ * @return Permissions as fs::perms
+ */
+std::optional<fs::perms> InterpretPermSting(const std::string& s);
```
nit:  spelling

```diff
- InterpretPermSting
+ InterpretPermString
```

willcl-ark replied:
fixed

tdb3 reviewed (approved):
re ACK 4f6b00c94b59c8206c77647c97f4ed188fa99401

Ran rpc_users locally (passed).
Re-ran the manual check in https://github.com/bitcoin/bitcoin/pull/28167#pullrequestreview-2105557076 (same results).
Also sanity checked on a Windows machine running Python 3.9.13 that `platform.system()` does return "Windows".
Left one minor nit.

achow101 commented:
ACK 73f0a6cbd0b628675028fbd5a37eff8115e7ccfe

ryanofsky reviewed (approved):
Code review ACK 73f0a6cbd0b628675028fbd5a37eff8115e7ccfe. Main change since last review is no longer throwing a skip exception in the rpc test on windows, so other checks can run after it, and overall test result is passing, not skipped. Also were clarifying renames and documentation improvements.

tdb3 reviewed (approved):
cr ACK 73f0a6cbd0b628675028fbd5a37eff8115e7ccfe

Sjors commented:
Nice! Will try this out soon.
